File: backend/mr-db-sql/src/main/java/io/featurehub/db/services/ArchiveStrategy.java
package io.featurehub.db.services;

import io.featurehub.db.model.DbApplication;
import io.featurehub.db.model.DbApplicationFeature;
import io.featurehub.db.model.DbEnvironment;
import io.featurehub.db.model.DbGroup;
import io.featurehub.db.model.DbOrganization;
import io.featurehub.db.model.DbPerson;
import io.featurehub.db.model.DbPortfolio;
import io.featurehub.db.model.DbServiceAccount;

public interface ArchiveStrategy {
  void archivePortfolio(DbPortfolio portfolio);
  void archiveApplication(DbApplication application);
  void archiveEnvironment(DbEnvironment environment);
  void archiveOrganization(DbOrganization organization);
  void archiveServiceAccount(DbServiceAccount serviceAccount);
  void archiveGroup(DbGroup group);
  void archiveApplicationFeature(DbApplicationFeature feature);
  void archivePerson(DbPerson person);
}

File: backend/mr-db-sql/src/main/kotlin/io/featurehub/db/services/ArchiveStrategy.kt
package io.featurehub.db.services

import io.featurehub.db.model.*
import java.time.format.DateTimeFormatter

interface ArchiveStrategy {
  companion object {
    val isoDate = DateTimeFormatter.ISO_DATE_TIME
  }

  fun archivePortfolio(portfolio: DbPortfolio)
  fun archiveApplication(application: DbApplication)
  fun archiveEnvironment(environment: DbEnvironment)

  /**
   * If you are interested in the archiving of an environment, register a callback
   */
  fun environmentArchiveListener(listener: (DbEnvironment) -> Unit)
  fun archiveOrganization(organization: DbOrganization)
  fun archiveServiceAccount(serviceAccount: DbServiceAccount)
  fun archiveGroup(group: DbGroup)
  fun archiveApplicationFeature(feature: DbApplicationFeature)
  fun featureListener(listener: (DbApplicationFeature) -> Unit)
  fun archivePerson(person: DbPerson)
  fun archiveApplicationStrategy(strategy: DbApplicationRolloutStrategy, personWhoArchived: DbPerson)
}
